### ISSUE TITLE
Add p.drawer, p.results, etc.

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#414]
**p.drawer** which would contain p's drawer text, containing only text attributes. This will be a python property, wrapping a string uA.  Similarly, **p.results** would be another python property, wrapping a uA containing any picklable value. 

This item contains lengthy discussions of how to show p.drawer or p.results.  It's moot for now. The post [ENB: about emacs drawers](https://groups.google.com/forum/#!topic/leo-editor/-IgepXKz2m4) explains why I think this issue is misguided.

As Terry [points out](https://groups.google.com/d/msg/leo-editor/JHFM-9vaeFM/DyH105eyAgAJ), the attrib_edit plugin already does something similar.

**Comment by tbnorth** [comment]
The timeline of title changes is amusing :-)

Bear in mind that when attrib_edit is in its own pane, it's always displaying the attribs. of the current node, without the need to open / close a popup.  Of course I'm still ignorant of org. mode drawers, just dubious that something requiring a popup can be smoother than something that doesn't.

**Comment by edreamleo** [comment]
On Sat, Sep 23, 2017 at 11:09 AM, Terry Brown <notifications@github.com>
wrote:

> The timeline of title changes is amusing :-)
>
​Yeah, I would like to be able to delete or hide such history.​

> Bear in mind that when attrib_edit is in its own pane, it's always
> displaying the attribs. of the current node, without the need to open /
> close a popup. Of course I'm still ignorant of org. mode drawers, just
> dubious that something requiring a popup can be smoother than something
> that doesn't.
>
​org-mode drawers <http://orgmode.org/manual/Drawers.html> are typically
hidden, but can be shown with the typical wonky Emacs key combo.  Imo, the
big advantage of the drawers is that they are actually text, just like
everything else in org-mode.


A
​t present, Leo's .org importer and exporter plugins know nothing about
:x:​ syntax, as I have just double checked, so drawers probably import and
export already.  But that does not mean that drawers "just work" in Leo! To
the contrary, execute-script will barf on the :x: syntax.


I suppose there are at least two approaches.  The first, implied by the
opening comment in this issue, would be to convert drawer syntax to uA's.
This completely hides drawers from execute-script, at the cost of requiring
some way of viewing *and changing* uA's corresponding to drawers.


The second would add explicit support for drawers to all kinds of body
text.  This would complicate execute-script, and perhaps other parts of
Leo, but would build into Leo quite a nice way of representing text uA's
into arbitrary nodes.


I don't have a strong opinion about which approach would be best.

Edward

**Comment by tbnorth** [comment]
Hmm, now that I've read the drawers docs. they do seem awkward in a Leo context. If you just want arbitrary sublevels, you'd use sub-nodes.  But if you want things *sometimes* visible in the main text pane, but sometimes hidden, the two Leonine solutions that spring to mind are wikiview that hides text until you enter it, and... well, i guess that's it.  I was going to say a "viewrendered" pane that optionally shows them, but that would only be an adjunct to a wikiview approach.

**Comment by edreamleo** [comment]
On Sat, Sep 23, 2017 at 5:46 PM, Terry Brown <notifications@github.com>
wrote:

> Hmm, now that I've read the drawers docs. they do seem awkward in a Leo
> context. If you just want arbitrary sublevels, you'd use sub-nodes. But if
> you want things *sometimes* visible in the main text pane, but sometimes
> hidden, the two Leonine solutions that spring to mind are wikiview that
> hides text until you enter it, and... well, i guess that's it. I was going
> to say a "viewrendered" pane that optionally shows them, but that would
> only be an adjunct to a wikiview approach.
>

​Thanks for this.  ​I'm not sure what exactly I have considered in the
past, but let's follow your wikiview idea.  The *effect* is, I think,
exactly what we want.  That is, we want a way to show/hide drawers in every
node.

But how to do this? In other words, is the drawer text "real" (always
present) text or not?  Either way presents difficulties.  If we use uA's to
represent drawers, then the show-drawer command must *insert* text into the
body pane.  If drawers are real text, then all of Leo must adjust to this
new text.

It's a conundrum.

Edward

**Comment by tbnorth** [comment]
I don't think this needs to work ubiquitously throughout Leo because I think Leo has more Leonine solutions to this need (subnodes, uA/attrib_edit).  I think it just needs to work for org. mode emulation.  I don't think org mode allows

```
:drawername:
yada yada
:end:
```

in the middle of a `.c` file, for example.

So it's just a matter of allowing wikiview to match and hide those patterns.  Currently I think it handles single line expressions, match wise.  It might also need to play with line spacing as it plays with font size, to hide multiple lines.

At least that's my take.  org mode emulation seems like a bridge to encourage Emacs users to try Leo, compared to say Jupyter integration, which is making Leo a part of a large analytical framework which integrates with a lot of high value technologies (Jupyter can integrate with distributed data and computing systems for example).  So I think the effort they deserve is proportional to those end points.

**Comment by edreamleo** [comment]
On Sun, Sep 24, 2017 at 10:01 AM, Terry Brown <notifications@github.com>
wrote:

> I don't think this needs to work ubiquitously throughout Leo because I
> think Leo has more Leonine solutions to this need (subnodes,
> uA/attrib_edit). I think it just needs to work for org. mode emulation. I
> don't think org mode allows
>
> :drawername:
> yada yada
> :end:
>
> in the middle of a .c file, for example.
>
​Good point!​

> So it's just a matter of allowing wikiview to match and hide those
> patterns. Currently I think it handles single line expressions, match wise.
> It might also need to play with line spacing as it plays with font size, to
> hide multiple lines.
>
​So are you saying Leo doesn't need uA's corresponding to drawers?  That
would be sweet.

> At least that's my take. org mode emulation seems like a bridge to
> encourage Emacs users to try Leo, compared to say Jupyter integration,
> which is making Leo a part of a large analytical framework which integrates
> with a lot of high value technologies (Jupyter can integrate with
> distributed data and computing systems for example). So I think the effort
> they deserve is proportional to those end points.
>
​Always a good policy.

Edward

**Comment by xgid** [comment]
Hi there!

I've been waiting for this issue with great interest from the beginning because I see it as an opportunity for Leo to make a step forward in some "usability" fronts, not just an "org-mode compatibility issue". 

So let me first explain some of the main *features* I recall for org-mode **Drawers** (although I may be missing some, because I'm not an **expert** in org-mode):

1. They allow adding metadata to an existing "node" (or "entry" as sometimes they call them)
1. Drawer IDs and contents are free-text forms, yet there are some "system" drawers with well defined IDs and contents (e.g. LOGBOOK or PROPERTIES)
1. They easily show you which _type of metadata_ is already there for a node (by showing their ID)
1. They hide the metadata details (values) from the "normal" view
1. They allow an easy (inline) way to see and **edit** each metadata

As pointed out by @edreamleo and @tbnorth, some of these *features* can already be emulated in Leo by using uAs, the attrib_edit or the wikiview plugin... but I currently don't see how none of these plugins (or any combination of) can provide **ALL** the features detailed above. Edward's idea of adding new headline icons tries also to cover some of the above *features*, but not fully. And I really think that *ALL* those features are equally important to make "metadata" info really *usable*.

What I want to stress with this comment is that, before we go on trying to **implement** org-mode Drawers in Leo, we should really define which are our *goals* or desired *features* to have in Leo. What do you think?

**Comment by xgid** [comment]
Uau! I have just seen [this comment in a recent Edward's post](https://groups.google.com/d/msg/leo-editor/ktFTJAE6QjU/koPOGcszAgAJ) proposing a similar approach:

> This is why I am proposing that we read the users manuals for other systems, looking for goals. Only after we have a clear idea of what we want does it make sense to look into how to do it.  This was an Aha that came to me on vacation.

So it seems we are in-line here! 😄

**Comment by edreamleo** [comment]
On Tue, Oct 24, 2017 at 11:34 PM, XGiD <notifications@github.com> wrote:

So let me first explain some of the main *features* I recall for org-mode
> *Drawers* (although I may be missing some, because I'm not an *expert* in
> org-mode):
>
>    1. They allow adding metadata to an existing "node" (or "entry" as
>    sometimes they call them)
>
> ​A uA implementation of drawers will do this directly.​

>
>    1. Drawer IDs and contents are free-text forms, yet there are some
>    "system" drawers with well defined IDs and contents (e.g. LOGBOOK or
>    PROPERTIES)
>
> ​A viewer for uA's will do this.  We can define exactly the same system
uA's as in org-mode, or any other special-purpose drawers we like.

>
>    1. They easily show you which *type of metadata* is already there for
>    a node (by showing their ID)
>
> ​A viewer for uA's will do this.​


>
>    1. They hide the metadata details (values) from the "normal" view
>
> ​uA icons could be toggled.

>
>    1. They allow an easy (inline) way to see and *edit* each metadata
>
> ​A viewer for uA's will do this.​


>
>
> As pointed out by @edreamleo <https://github.com/edreamleo> and @tbnorth
> <https://github.com/tbnorth>, some of these *features* can already be
> emulated in Leo by using uAs, the attrib_edit or the wikiview plugin... but
> I currently don't see how none of these plugins (or any combination of) can
> provide *ALL* the features detailed above.
>
​Imo, the wikiview plugin is a desirable feature, but not for this
discussion.  That is, implementing drawers as uA's explicitly precludes
plain-text drawers in nodes.  Drawers should be, as you say, separate
meta-data attached to nodes.

> Edward's idea of adding new headline icons tries also to cover some of the
> above *features*, but not fully.
>
​Headline icons merely indicate which nodes have general uA's or specific
drawer-related uA's.  They are completely separate from all other aspects
of the proposal.​


> And I really think that *ALL* those features are equally important to
> make "metadata" info really *usable*.
>
​So do I.  As shown above, uA's, combined with an easy-to-use viewer, will
give ​all these features directly and naturally.

> What I want to stress with this comment is that, before we go on trying to
> *implement* org-mode Drawers in Leo, we should really define which are
> our *goals* or desired *features* to have in Leo. What do you think?
>
​The goals are easily stated:

1. To provide a text representation of drawers.  This is the text shown by
the uA viewer, which will allow users to see, change, insert and delete
drawers.

2. To allow scripts easy access to the data in drawers.  This will be
provided by an api for drawers.​

Edward

**Comment by xgid** [comment]
> Headline icons merely indicate which nodes have general uA's or specific drawer-related uA's. They are completely separate from all other aspects of the proposal.​

**Comment by xgid** [comment]
> ​Imo, the wikiview plugin is a desirable feature, but not for this discussion. That is, implementing drawers as uA's explicitly precludes plain-text drawers in nodes. Drawers should be, as you say, separate meta-data attached to nodes.

> As shown above, uA's, combined with an easy-to-use viewer, will give ​all these features directly and naturally.

This makes your general proposal more clear for me now. A key factor is the "easy-to-use viewer"! I have some ideas in this regard, which I don't have the time to expose at this moment, but let me just say that I don't see neither the pop-up window nor the currently existing attrib_edit plugin editor as really *easy-to-use* in this context. I'll try to explain my ideas later.

> Headline icons merely indicate which nodes have general uA's or specific drawer-related uA's. They are completely separate from all other aspects of the proposal.​

OK, but I still don't see how are you thinking in providing the feature of "_easily show you which type of metadata is already there for a node (by showing their ID)_" in a way that these IDs are explicitly shown in the Leo tree in some way. I understand that it cannot be done with headline icons alone.

**Comment by tbnorth** [comment]
Not an org mode user, but I got the impression the interleaving of
drawers with other text part of how they worked:
http://orgmode.org/manual/Drawers.html

Maybe that's unimportant in practice, but if it does matter (to org
mode users?) I'm not sure the uA approach covers it.

This would be exacerbated if multiple drawers with the same name are
allowed, don't know about that either.

**Comment by edreamleo** [comment]
On Wed, Oct 25, 2017 at 7:58 AM, XGiD <notifications@github.com> wrote:

I don't see neither the pop-up window nor the currently existing
> attrib_edit plugin editor as really *easy-to-use* in this context. I'll
> try to explain my ideas later.


​Imo, this is a minor issue.  Leo already has an api for dealing with uA's
and will have a simpler api for dealing with drawers.  In practice, I
assume that modifying drawers by hand will be rare.

> Headline icons merely indicate which nodes have general uA's or specific
> drawer-related uA's. They are completely separate from all other aspects of
> the proposal.​
>
> OK, but I still don't see how are you thinking in providing the feature of
> "*easily show you which type of metadata is already there for a node (by
> showing their ID)*"
>
Showing the drawer visually should suffice for users.  The drawer's api
will suffice for scripts.

Edward

**Comment by edreamleo** [comment]
On Wed, Oct 25, 2017 at 8:41 AM, Terry Brown <notifications@github.com>
wrote:

> Not an org mode user, but I got the impression the interleaving of
> drawers with other text part of how they worked:
> http://orgmode.org/manual/Drawers.html
>
> Maybe that's unimportant in practice, but if it does matter (to org
> mode users?) I'm not sure the uA approach covers it.
>

​In org mode *everything* is text, not just drawers.  Leo's world is far,
far richer than this.  Imo, the only thing that matters is that the
org-mode writer writes the uA's as text.  That will provide compatibility
within org mode itself.  Within Leo, the drawer's api will be associated
not with plain text, but with a Leo node, a real live python object.

Edward

**Comment by edreamleo** [comment]
On Wed, Oct 25, 2017 at 8:41 AM, Terry Brown <notifications@github.com>
wrote:

Not an org mode user, but I got the impression the interleaving of
> drawers with other text part of how they worked:
>

​By analogy with @doc, we could contemplate adding support for @drawer
<drawer name>.  This would continue until the next @doc, @drawer, @c or the
end of the body text.​

We agree that support for org mode should not "infect" the rest of Leo.
But @drawer would affect the code that reads and writes files, as well as
all code that executes scripts.

As a result, I rejected this idea for compatibility reasons.  Older
versions of Leo would not handle @drawer properly.

Let's not forget a big big difference between Leo and org mode.  In Leo,
*all* body text may be a script. In org mode, all scripts must be delimited
by the hideous #+SRC syntax. (Org mode also lacks @others.)

In Leo, the @doc (or @) syntax delimits non-script from script. In the
ancient notion of literate programming, such comments were to be
interleaved with text.  The Leonine version of LP dispenses with most such
comments, replacing them with outline structure or regular comments.

So even if we added @drawer to Leo, huge differences would remain between
Leo and org mode.  Imo, it is futile and unnecessary to attempt exact
equivalence between org mode and Leonine features.  All that matters is
that org mode files round trip between Emacs and Leo in a straightforward
way.  This is good news, because it makes our job straightforward.

Edward

**Comment by xgid** [comment]
> Not an org mode user, but I got the impression the interleaving of drawers with other text part of how they worked: http://orgmode.org/manual/Drawers.html 
> Maybe that's unimportant in practice, but if it does matter (to org mode users?) I'm not sure the uA approach covers it.

Yes, _in theory_ (according to [org mode syntax](https://github.com/izahn/worg-mirror/blob/master/dev/org-syntax.org#drawers-and-property-drawers)) they *can* be interleaved with text, but _in practice_ it seems that they are currently **ONLY** being used right under the **headline** or under a [**PLANNING** element](https://github.com/izahn/worg-mirror/blob/master/dev/org-syntax.org#clock-diary-sexp-and-planning), if present. 

I came up to this conclusion after:

* Looking at **all** current uses of drawers at the [Org community-driven documentation and code snippets](https://github.com/izahn/worg-mirror) source code, and
* Searching for any other uses of drawers at any Org-mode file in any public GitHub repository

In all of them, the mainly used drawer is the `:PROPERTIES:` one, with some sporadic use of `:LOGBOOK:`. The only "caveat" is that sometimes both drawers are used together in the same node: PROPERTIES first, immediately followed by LOGBOOK.

> This would be exacerbated if multiple drawers with the same name are allowed, don't know about that either.

It seems they are not, at least _in practice_.

So I would not try to complicate our implementation to cope with functionality that nobody seems to be using at present.

**Comment by xgid** [comment]
OK, I'll try to explain now my point of view regarding the use and edition of uAs in Leo, which can also be applied to drawers, but is not exclusive to them. In fact, I'm more concerced with other current uses of uAs by different Leo plugins like todo.py, nodetags.py, attrib_edit.py and backlink.py.

There are two different aspects of the question which are to me equally important:

1. The visibility of **the existance** and **the data** contained in the uAs
2. The **easy** edition of those uAs

The Todo plugin is a really good example of both aspects: it adds some icons to the node to indicate the **existance** of some attributes (like the "duedate") or to show the **value** of others (like the "priority"). And it even shows a summary in the headline (and "automatically" updates it) for some other data like the task completion rate:
![imagen](https://user-images.githubusercontent.com/12764559/32063940-dfe670f4-ba4e-11e7-9b1c-93d95624e7c4.png)

All those attributes are stored as uAs for each node, so they are normally hidden from the main tree view, but the plugin handles what information must be made visible in each node and how.

And then, there's a nice editor which allows you to view the full contents of those and the other attributes and to modify each attribute with the best UI for each case.

But, what if I want to see the due date or the next working date of each node also summarised in the headline, like Emacs Org-mode does?

What if I want to see the tags currently present in a node also in the headline, like Emacs Org-mode does?

And finally, the main problem comes when I use all those plugins together in the same document: I have nodes with Todo properties, others with tags, others with backlinks **and others with all of them in the same node**. With the current Leo approach, I have to jump from **one tab to another in the log pane** to see what data is present in the node and to edit it. And making a stack of tabs using the free_layout system to show all of them all the time is not feasible as you can imagine.

A better approach would be to have a way to automatically show the desired plugin(s) editor(s) when entering a node which has attributes of that plugin. Or, at least, stack them up when you invoke an appropriate command (which can be bound to a shortcut) and hide them again when I leave the node. Maybe it can be done with the currenty Leo API and I'm not aware?

Of course, it would be even better to just be able to double-click the desired area of the headline (being it the priority icon, the duedate icon, the progress summary or the tags summary) and jump straight to the appropriate editor! Or even show the editor inline! But I understand that these other approaches are currently totally out of Leo's capabilities, aren't they?

As you see, this comment is only tangentially related with Org-mode and its drawers, but they were the triggers of this ideas for me so I wanted to share them here. I can move this debate to another issue or to the Google group if you see fit.

**Comment by tbnorth** [comment]
Interesting post, thank you.  Comments in-line below.

On Thu, 26 Oct 2017 16:57:42 +0000 (UTC)
XGiD <notifications@github.com> wrote:

> OK, I'll try to explain now my point of view regarding the use and
> edition of uAs in Leo, which can also be applied to drawers, but is
> not exclusive to them. In fact, I'm more concerced with other current
> uses of uAs by different Leo plugins like todo.py, nodetags.py,
> attrib_edit.py and backlink.py.
> 
> There are two different aspects of the question which are to me
> equally important:
> 
> 1. The visibility of **the existance** and **the data** contained in
> the uAs 2. The **easy** edition of those uAs
> 
> The Todo plugin is a really good example of both aspects: it adds
> some icons to the node to indicate the **existance** of some
> attributes (like the "duedate") or to show the **value** of others
> (like the "priority"). And it even shows a summary in the headline
> (and "automatically" updates it) for some other data like the task
> completion
> rate: ![imagen](https://user-images.githubusercontent.com/12764559/32063940-dfe670f4-ba4e-11e7-9b1c-93d95624e7c4.png)
> 
> All those attributes are stored as uAs for each node, so they are
> normally hidden from the main tree view, but the plugin handles what
> information must be made visible in each node and how.
> 
> And then, there's a nice editor which allows you to view the full
> contents of those and the other attributes and to modify each
> attribute with the best UI for each case.
> 
> But, what if I want to see the due date or the next working date of
> each node also summarised in the headline, like Emacs Org-mode does?
> 
> What if I want to see the tags currently present in a node also in
> the headline, like Emacs Org-mode does?

At first(*) I was going to say that these features would require extra
code, and if you're writing extra code, discovering the layout of each
plugin's data in uA isn't hard.  To be clear, if you can write extra
code to place transient information in the headline, then you must know
enough Python to be able to inspect uA.  Perhaps I misunderstand your
point, but I guess I'm saying the barrier to adding that info. to the
headlines is not the use of uAs, but the need to write extra code.

Perhaps you mean that an org mode like approach would have put all that
info. in text in the first place, avoiding the whole problem of extra
code or uAs.  That may be true, but Leo is not a plain text
environment.  There's a limit to how much you can get in a headline,
and Leo makes it possible to make headlines somewhat dynamic, which I
think is a plus, but not as simple as plain text.  Again I'm not 100%
sure I'm interpreting your point correctly, but that's what I get from
this.
 
> And finally, the main problem comes when I use all those plugins
> together in the same document: I have nodes with Todo properties,
> others with tags, others with backlinks **and others with all of them
> in the same node**. With the current Leo approach, I have to jump
> from **one tab to another in the log pane** to see what data is
> present in the node and to edit it. And making a stack of tabs using
> the free_layout system to show all of them all the time is not
> feasible as you can imagine.

Just to clarify, do you mean not feasible because you don't have enough
screen space, or because free_layout is too cumbersome?  I've certainly
used layouts displaying three or more "plugin editor" panes.

> A better approach would be to have a way to automatically show the
> desired plugin(s) editor(s) when entering a node which has attributes
> of that plugin. Or, at least, stack them up when you invoke an
> appropriate command (which can be bound to a shortcut) and hide them
> again when I leave the node. Maybe it can be done with the currenty
> Leo API and I'm not aware?

These are great ideas which I would like to use in free_layout's
eventual replacement, the QDocks layout.  Personally I'd find panes
popping open and closed as I navigate around distracting generally, but
it could be a great space saver on laptop displays etc.

> Of course, it would be even better to just be able to double-click
> the desired area of the headline (being it the priority icon, the
> duedate icon, the progress summary or the tags summary) and jump
> straight to the appropriate editor! Or even show the editor inline!
> But I understand that these other approaches are currently totally
> out of Leo's capabilities, aren't they?

Yep - inline editors in the tree widget etc. are possible but I don't
see anyone wanting to work on such things anytime soon - I think the
selective display of appropriate editor panes has a lot more immediate
appeal.

I've made note of these ideas here:
https://github.com/orgs/leo-editor/projects/2

Cheers -Terry

> As you see, this comment is only tangentially related with Org-mode
> and its drawers, but they were the triggers of this ideas for me so I
> wanted to share them here. I can move this debate to another issue or
> to the Google group if you see fit.

**Comment by xgid** [comment]
Thanks for your detailed answer, Terry. I'll try to clarify (inline).

> At first(*) I was going to say that these features would require extra
> code, and if you're writing extra code, discovering the layout of each
> plugin's data in uA isn't hard.  To be clear, if you can write extra
> code to place transient information in the headline, then you must know
> enough Python to be able to inspect uA.  Perhaps I misunderstand your
> point, but I guess I'm saying the barrier to adding that info. to the
> headlines is not the use of uAs, but the need to write extra code.
> 
> Perhaps you mean that an org mode like approach would have put all that
> info. in text in the first place, avoiding the whole problem of extra
> code or uAs.  That may be true, but Leo is not a plain text
> environment.  There's a limit to how much you can get in a headline,
> and Leo makes it possible to make headlines somewhat dynamic, which I
> think is a plus, but not as simple as plain text.  Again I'm not 100%
> sure I'm interpreting your point correctly, but that's what I get from
> this.

Now I see that all my comments may be interpreted as:

1. A critique to the use of uAs in Leo
2. A claim for the use of plain text "à la" Emacs Org mode
3. A critique to the complexity of doing some things in Leo

So let's make my point of view clear regarding those statements because they are not true "as is":

1. I really see Leo's uAs **as the way to go to store any metadata attached to a node**, because they allow structuring the information as you see fit and so the possibility of a powerful API to access it. The use of uAs is not "the problem" I'm trying to stress here, but the way that uAs are currently **visualized** and **edited** in Leo. Maybe this has to do **in part** with the fact that uAs are mainly created and managed directly by plugins (which are independent one from the other), not by Leo core, so their visualization and edition are more difficult to "coordinate" in a unified way. But I also think that the use of pugins is correct: by saying this, I don't want to suggest that I'm against them! Maybe is a question of "coordination" between different plugins.
As a side note: uAs could be enhanced with more database-like properties to give them still more power, but that's another question that we can leave for another moment.
2. Plain text cannot be as useful as structured info, as Leo has demonstrated! The problem is that the use of uAs in Leo takes that information "far away" from user's view. So far that you **always** need specialised code to see it. In this regard, Emacs Org mode has a win: all metadata is just a "Tab" key away. I would really like to also see that in Leo. What about a Unified Editor of uAs?! 
The attrib_edit plugin does not help here because:
	* It can only edit "their own attributes" (those with `_edit` in the path)
	* It does not know how to show you the best editor for the attribute at hand
3. One of Leo's strong points is the possibility of doing nearly **anything you want** programmatically, that's true. But that also makes the learning curve quite flat... :). In this regard, the "everything is plain text" paradigm in Org mode has already shown that it does NOT make thinks easier for the programmer... so that's not the point. I think that Leo is clearly a winner here. But the point is that "common things" should have "common ways" to do them, for the benefit of the users. And visualization and edition of metadata in nodes (using uAs in Leo) is already a "common thing".

To summarise, My point in this regard related to uAs is that: as there is a common pattern 

> Just to clarify, do you mean not feasible because you don't have enough
> screen space, or because free_layout is too cumbersome?  I've certainly
> used layouts displaying three or more "plugin editor" panes.

The problem is not the use of free_layout. I've done it:

![imagen](https://user-images.githubusercontent.com/12764559/32081940-abab45e0-ba8e-11e7-9d7c-0b268a1c4097.png)

but it simply does not scale! And it's far from _usable_ also (information is difficult to read, etc.).

> These are great ideas which I would like to use in free_layout's
> eventual replacement, the QDocks layout.  Personally I'd find panes
> popping open and closed as I navigate around distracting generally, but
> it could be a great space saver on laptop displays etc.

As stated above, maybe the "solution" is not popping panes up and down. Maybe its a unified uA's pane with sections for each plugin. Or maybe an inline editing of as much uAs as possible and only limited panes for the rest. For example, a common pattern nowadays for tags is to show them inline, as in GitHub:

![imagen](https://user-images.githubusercontent.com/12764559/32081949-b4884e1a-ba8e-11e7-865d-409d8206e1c4.png)

or Workflowy (which even let's you edit them inline à la Org mode):
![imagen](https://user-images.githubusercontent.com/12764559/32081953-b7f7ba4a-ba8e-11e7-8a17-fa62b84b8502.png)

And in both cases, a simple click in a tag adds it to the search box to filter all entries by it.

That would make the tag pane nearly unnecessary.

> Yep - inline editors in the tree widget etc. are possible but I don't
> see anyone wanting to work on such things anytime soon - I think the
> selective display of appropriate editor panes has a lot more immediate
> appeal.

Yes, I understantd that. Anyway, let me insist that an "inline editor" in some cases, tags for example, could be as simple as a plain-text editor if you want a quick edition. You can always go for the advanced editor if you need it. It would probably require the use of some kind of "text mark" (as '#') to indicate that you are introducing a tag and not other text or something like that. Specific enhancements to the tags plugin may be better treated as separate issues.

> I've made note of these ideas here:
> https://github.com/orgs/leo-editor/projects/2

The link is not working for me at this moment. And the [projects page](https://github.com/leo-editor/leo-editor/projects) in leo-editor shows only the "Jupyter Integration" project created by Edward.

**Comment by xgid** [comment]
To summarise, my point in this regard related to uAs is that: as there is a common pattern...
... there could be a common approach to address it:

* There's the need to have metadata attached to nodes in the uAs structured way (and set apart from the main node contents), but at the same time the need for easy ways to show this metadata **inline** in the node or **near to it**. This is what the Todo plugin does quite well, but the tags plugin does not, for example.
* There's the need to search for nodes with specific metadata attributes or specific attribute values. This is currently done in **different and separate** ways by each plugin and cannot be combined (without custom code, of course!).
* There's the need to edit each attribute with an appropriate editor according to its "type" (dates, integers, strings, etc.). Again, each plugin provides its own editor in separate panes.

I think this could be done in a more "unified" way. And now that we are talking about adding yet another uA (drawers), it's a good momento to talk about it, don't you think?

**Comment by xgid** [comment]
Following recent Edward's recommendation to keep issues short (which I agree), we should probably just address the **p.drawers** issue here and take other enhancements to other issues. 

Tell me if you need any help from me to do it.

**Comment by edreamleo** [comment]
I am going to close this item and label it "Won'tDo". The post [ENB: about emacs drawers](https://groups.google.com/forum/#!topic/leo-editor/-IgepXKz2m4) explains why I think this issue is misguided.
